### PR TITLE
Dc serial no device object

### PR DESCRIPTION
**Pull request by dcmaverick**
Adding Serial No to device object

### CODE REVIEW

File: src/bacnet/basic/object/device.c [hunk 1]
```diff
@@ -471,7 +471,7 @@ static const int Device_Properties_Required[] = { PROP_OBJECT_IDENTIFIER,
     PROP_PROTOCOL_OBJECT_TYPES_SUPPORTED, PROP_OBJECT_LIST,
     PROP_MAX_APDU_LENGTH_ACCEPTED, PROP_SEGMENTATION_SUPPORTED,
     PROP_APDU_TIMEOUT, PROP_NUMBER_OF_APDU_RETRIES, PROP_DEVICE_ADDRESS_BINDING,
-    PROP_DATABASE_REVISION, -1 };
+    PROP_DATABASE_REVISION,PROP_SERIAL_NUMBER, -1 };
```
Comment (metelik) [inline src/bacnet/basic/object/device.c:474]: space after comma

File: src/bacnet/basic/object/device.c [hunk 2]
```diff
@@ -515,6 +515,7 @@ static BACNET_DEVICE_STATUS System_Status = STATUS_OPERATIONAL;
 static char *Vendor_Name = BACNET_VENDOR_NAME;
 static uint16_t Vendor_Identifier = BACNET_VENDOR_ID;
 static char Model_Name[MAX_DEV_MOD_LEN + 1] = "GNU";
+static char Serial_Number[MAX_DEV_MOD_LEN + 1];
```
Comment (metelik) [inline src/bacnet/basic/object/device.c:518]: initialise it to "" please

File: src/bacnet/basic/object/device.c [hunk 3]
```diff
@@ -844,6 +845,20 @@ bool Device_Set_Model_Name(const char *name, size_t length)
     return status;
 }
 
+bool Device_Set_Serial_Number(const char *serial_number, size_t length)
+{
+    bool status = false; /*return value */
+
+    if (length < sizeof(Serial_Number)) {
+        memmove(Serial_Number, serial_number, length);
```
Comment (metelik) [inline src/bacnet/basic/object/device.c:853]: strncpy instead of memmove?

File: src/bacnet/basic/object/device.c [hunk 4]
```diff
@@ -844,6 +845,20 @@ bool Device_Set_Model_Name(const char *name, size_t length)
...
+        Serial_Number[length] = 0;
```
Comment (metelik) [inline src/bacnet/basic/object/device.c:854]: = '\0' is more proper ;)

**Review by metelik — APPROVED** [review]
Few code improvements suggested...

**Review by PeterJeaiem — APPROVED** [review]

File: src/bacnet/basic/object/device.c (hunk 3 above)
Comment (dcmaverick) [reply src/bacnet/basic/object/device.c:853]: memmove is what is used in the other functions in this file, so I was just following its lead for compatibility reasons

File: src/bacnet/basic/object/device.c (hunk 2 above)
Comment (adrigonzo) [reply src/bacnet/basic/object/device.c:518]: We can probably use the same default as we use in our Web UI, "0000"

**Review by adrigonzo — APPROVED** [review]

File: src/bacnet/basic/object/device.c (hunk 2 above)
Comment (metelik) [reply src/bacnet/basic/object/device.c]: Well.. Each and every product will have prolly some other default which we do not know and hence cannot accommodate upfront.

**Pull request merged**